File: packages/sdk-py/dna/kernel/studio_ui.py
"""StudioUIMetadata — UI declarations attached to KindPort.

Phase 2 of the route-manifest Big-Bang (s-studio-ui-metadata, 2026-05-16).

Anchors Studio UI metadata (mode, routes, labels, RBAC, i18n) in the
DNA SDK itself. Studio fetches `GET /kinds/manifest` at boot, merges
with its static TS-side manifest, and generates router + sidebar +
sitemap entries automatically.

Why anchor in the SDK?
  - Third-party extensions ship Kinds + UI metadata together. No
    Studio code change needed when a new Kind appears.
  - RBAC declared alongside the Kind (single source instead of
    Python KIND_WRITE_ROLES + TS requiredAnyRole duplicate).
  - i18n: labels can be a {locale: str} dict — Studio picks the
    user's locale at boot.
  - Schema-driven CRUD (Phase 3+): the same Kind.schema() that
    validates writes also drives form rendering.

Compatibility: `ui` is OPTIONAL. Existing ~30 Kinds without `ui`
populated remain invisible in Studio's manifest-driven surfaces.
The Phase 1 TS-side route-manifest.ts still defines app-shell +
mode-landing + workflow routes that aren't Kind-specific.

Backwards compat: schema is forward-compatible. New optional fields
can be added without breaking existing Kinds. Old Studio versions
ignore fields they don't understand (the harness response includes
``manifest_version`` for explicit checks).
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Literal


# ─────────────────────────────────────────────────────────────────────
# Type aliases
# ─────────────────────────────────────────────────────────────────────

ModeId = Literal["plan", "build", "quality", "govern", "cognitive"]
"""Studio mode this Kind primarily belongs to."""

UIAction = Literal["list", "detail", "edit", "create"]
"""CRUD action a route serves."""

LabelI18n = dict[str, str]
"""Locale → translated string. e.g. {'en': 'Eval Cases', 'pt-BR': 'Casos'}."""


# ─────────────────────────────────────────────────────────────────────
# StudioUIMetadata
# ─────────────────────────────────────────────────────────────────────

@dataclass
class StudioUIMetadata:
    """UI declarations for a Kind, consumed by Studio's manifest.

    Populate on a KindPort by setting `ui = StudioUIMetadata(...)`.
    Defaults are conservative: no mode, no sidebar, no routes — the
    Kind is invisible in Studio unless explicitly opted in.

    Example::

        class EvalExperimentKind(KindBase):
            kind = "EvalExperiment"
            alias = "eval-evalexperiment"
            ...
            ui = StudioUIMetadata(
                mode="quality",
                label={"en": "Experiments", "pt-BR": "Experimentos"},
                icon="🔬",
                in_sidebar=True,
                routes={
                    "list":   "eval/experiments",
                    "detail": "eval/experiments/:expId",
                },
                permissions={
                    "list":   "any",
                    "detail": "any",
                    "create": ["maker", "qa"],
                    "edit":   ["maker", "qa"],
                },
                breadcrumb=["Quality", "Experiments"],
                display_order=20,
            )
    """

    # ── Mode + sidebar placement ─────────────────────────────────────

    mode: ModeId | None = None
    """Studio mode this Kind belongs to. None = the Kind exists in
    the system but doesn't claim a mode (won't appear in mode
    sidebars even with in_sidebar=True)."""

    in_sidebar: bool = False
    """When True, the Kind's list route renders in the mode's Sidebar
    via the manifest. Detail/edit/create routes never go in sidebar."""

    display_order: int = 100
    """Sort order in the Sidebar (lower = higher). Cluster items
    semantically: e.g. 10–19 for primary, 20–29 for secondary."""

    # ── Labels (i18n-ready) ──────────────────────────────────────────

    label: str | LabelI18n | None = None
    """Display label. Either a plain string OR a {locale: str} dict.
    None falls back to KindPort.display_label (legacy) or Kind name."""

    icon: str | None = None
    """Emoji or icon identifier. None falls back to ascii_icon."""

    description: str | LabelI18n | None = None
    """Longer description for tooltips + sitemap. i18n-ready."""

    breadcrumb: list[str] | None = None
    """Breadcrumb fragments from app root to this Kind's list route.
    E.g. ['Quality', 'Lab', 'Cases'] for EvalCase."""

    # ── Routes (per CRUD action) ─────────────────────────────────────

    routes: dict[UIAction, str] = field(default_factory=dict)
    """Path pattern per action. Leading "/" = ABSOLUTE (root-level);
    no leading slash = scope-relative (prefix /scopes/:scope/).

    Examples::

        routes={
            "list":   "docs/Story",                        # /scopes/X/docs/Story
            "detail": "docs/Story/:name",                  # /scopes/X/docs/Story/foo
            "create": "kinds/Story/__new__",
        }
    """

    # ── RBAC ─────────────────────────────────────────────────────────

    permissions: dict[UIAction, list[str] | str] = field(default_factory=dict)
    """Per-action role requirements. Each value is either:

      - ``"any"`` — any logged-in user
      - ``"anonymous"`` — no auth required (rare, e.g. public Doc)
      - ``list[str]`` — at least one role match required;
        power-user bypasses every gate

    Missing actions default to ``"any"`` for list/detail and to
    ``["architect", "maker"]`` for create/edit — sensible defaults
    that match the Phase 1 KIND_WRITE_ROLES baseline.
    """

    # ── Future extensions ─────────────────────────────────────────────

    note: str | None = None
    """Free-form annotation — surfaces in /studio/sitemap admin page."""

    feature_flag: str | None = None
    """Optional env-flag gating (e.g. 'VITE_STUDIO_COGNITIVE').
    Studio hides the Kind in UI when the flag isn't set. Backend
    routes still work — this is purely a visibility hint."""

    # ─────────────────────────────────────────────────────────────────

    def to_dict(self) -> dict[str, Any]:
        """Serialize for the /kinds/manifest JSON response.

        Omits None/empty fields so the wire payload stays small.
        """
        out: dict[str, Any] = {}
        if self.mode is not None:
            out["mode"] = self.mode
        if self.in_sidebar:
            out["in_sidebar"] = True
        if self.display_order != 100:
            out["display_order"] = self.display_order
        if self.label is not None:
            out["label"] = self.label
        if self.icon is not None:
            out["icon"] = self.icon
        if self.description is not None:
            out["description"] = self.description
        if self.breadcrumb is not None:
            out["breadcrumb"] = list(self.breadcrumb)
        if self.routes:
            out["routes"] = dict(self.routes)
        if self.permissions:
            out["permissions"] = {
                k: list(v) if isinstance(v, list) else v
                for k, v in self.permissions.items()
            }
        if self.note is not None:
            out["note"] = self.note
        if self.feature_flag is not None:
            out["feature_flag"] = self.feature_flag
        return out

    def resolve_label(self, locale: str = "en") -> str | None:
        """Resolve i18n label for the requested locale.

        Fallback order:
          1. Exact locale match (e.g. 'pt-BR').
          2. Language-only fallback (e.g. 'pt' from 'pt-BR').
          3. 'en' (project default).
          4. None.

        Plain-string label is returned as-is regardless of locale.
        """
        if self.label is None:
            return None
        if isinstance(self.label, str):
            return self.label
        if locale in self.label:
            return self.label[locale]
        if "-" in locale:
            base = locale.split("-", 1)[0]
            if base in self.label:
                return self.label[base]
        return self.label.get("en")


# ─────────────────────────────────────────────────────────────────────
# docs_ui — convention helper (s-kinds-studio-ui-exposure)
# ─────────────────────────────────────────────────────────────────────

def docs_ui(
    kind: str,
    *,
    mode: ModeId,
    label_en: str,
    label_pt: str,
    icon: str | None = None,
    display_order: int = 60,
    description_en: str | None = None,
    description_pt: str | None = None,
    breadcrumb: list[str] | None = None,
    in_sidebar: bool = True,
    read: list[str] | str = "any",
) -> StudioUIMetadata:
    """Build a StudioUIMetadata for a Kind on Studio's GENERIC docs surface.

    Cuts the per-Kind boilerplate to a single call: routes are the universal
    scope-relative docs routes (``docs/<Kind>`` list + ``docs/<Kind>/:name``
    detail) — the SAME shape AuditLog uses, so they resolve for ANY Kind
    without bespoke Studio routing (no 404 risk). Read-only surface (list +
    detail); create/edit routes are intentionally omitted — those need
    Kind-specific forms and can be added per-Kind later.

    i18n labels are built from ``label_en`` + ``label_pt`` (Studio resolves the
    user's locale at boot). ``icon`` should usually pass the Kind's
    ``ascii_icon`` so the manifest matches the tree view.
    """
    label: LabelI18n = {"en": label_en, "pt-BR": label_pt}
    description: LabelI18n | None = None
    if description_en is not None:
        description = {"en": description_en, "pt-BR": description_pt or description_en}
    return StudioUIMetadata(
        mode=mode,
        in_sidebar=in_sidebar,
        display_order=display_order,
        label=label,
        icon=icon,
        description=description,
        breadcrumb=breadcrumb,
        routes={"list": f"docs/{kind}", "detail": f"docs/{kind}/:name"},
        permissions={"list": read, "detail": read},
    )

File: packages/sdk-ts/src/kernel/studio_ui.ts
/**
 * StudioUIMetadata — UI declarations attached to a KindPort.
 *
 * Net-new TS twin of Python `dna.kernel.studio_ui.StudioUIMetadata`
 * (studio_ui.py:54-215), added for the descriptor-expressiveness `ui:` field
 * (spec 2026-06-11 D1). Reverses the previously-documented posture that `ui`
 * was "a Studio-backend concern… intentionally not mirrored"
 * (src/extensions/audit.ts) — byte-parity of `.kind.yaml` descriptors across
 * runtimes requires BOTH the Python and the TS DeclarativeKindPort to parse the
 * same `ui:` block into the same data and serialize it identically.
 *
 * Field-for-field mirror + `toDict()` (same omission rules as Python
 * `to_dict()`) + `resolveLabel()` (same i18n fallback order as Python
 * `resolve_label()`). There is no TS class golden for `ui` (TS Kind classes
 * never carried it); the parity pin is structural — `toDict()` parity with
 * Python via the shared fixture tests/fixtures/studio-ui-parity.json.
 */

/** Studio mode this Kind primarily belongs to. */
export type ModeId = "plan" | "build" | "quality" | "govern" | "cognitive";

/** CRUD action a route serves. */
export type UIAction = "list" | "detail" | "edit" | "create";

/** Locale → translated string. e.g. {en: 'Eval Cases', 'pt-BR': 'Casos'}. */
export type LabelI18n = Record<string, string>;

/**
 * Canonical, ordered set of StudioUIMetadata field names. Single source of
 * truth — `KindDefinitionSpec.from_raw`/zod validate `ui:` keys against THIS
 * list (no second hardcoded list). Mirrors the Python dataclass field order.
 */
export const UI_METADATA_FIELDS = [
  "mode",
  "in_sidebar",
  "display_order",
  "label",
  "icon",
  "description",
  "breadcrumb",
  "routes",
  "permissions",
  "note",
  "feature_flag",
] as const;

export type UIMetadataField = (typeof UI_METADATA_FIELDS)[number];

/** Constructor input — every field optional, mirroring the Python dataclass
 *  defaults (no mode, no sidebar, display_order 100, all others None/empty). */
export interface StudioUIMetadataInit {
  mode?: ModeId | null;
  in_sidebar?: boolean;
  display_order?: number;
  label?: string | LabelI18n | null;
  icon?: string | null;
  description?: string | LabelI18n | null;
  breadcrumb?: string[] | null;
  routes?: Record<string, string>;
  permissions?: Record<string, string[] | string>;
  note?: string | null;
  feature_flag?: string | null;
}

/**
 * UI declarations for a Kind, consumed by Studio's manifest. Defaults are
 * conservative: no mode, no sidebar, no routes — the Kind is invisible in
 * Studio unless explicitly opted in. 1:1 with Python `StudioUIMetadata`.
 */
export class StudioUIMetadata {
  readonly mode: ModeId | null;
  readonly in_sidebar: boolean;
  readonly display_order: number;
  readonly label: string | LabelI18n | null;
  readonly icon: string | null;
  readonly description: string | LabelI18n | null;
  readonly breadcrumb: string[] | null;
  readonly routes: Record<string, string>;
  readonly permissions: Record<string, string[] | string>;
  readonly note: string | null;
  readonly feature_flag: string | null;

  constructor(init: StudioUIMetadataInit = {}) {
    this.mode = init.mode ?? null;
    this.in_sidebar = init.in_sidebar ?? false;
    this.display_order = init.display_order ?? 100;
    this.label = init.label ?? null;
    this.icon = init.icon ?? null;
    this.description = init.description ?? null;
    this.breadcrumb = init.breadcrumb ?? null;
    this.routes = init.routes ?? {};
    this.permissions = init.permissions ?? {};
    this.note = init.note ?? null;
    this.feature_flag = init.feature_flag ?? null;
  }

  /** The canonical field-name set (single source of truth for `ui:` validation). */
  static fields(): readonly UIMetadataField[] {
    return UI_METADATA_FIELDS;
  }

  /**
   * Serialize for the /kinds/manifest JSON response. Omits None/empty fields
   * so the wire payload stays small — EXACTLY the Python `to_dict()` omission
   * rules (studio_ui.py:161-192):
   *   - mode: only if not null
   *   - in_sidebar: only if true
   *   - display_order: only if !== 100
   *   - label/icon/description/note/feature_flag: only if not null
   *   - breadcrumb: only if not null (an empty list IS kept, like Python)
   *   - routes/permissions: only if non-empty
   */
  toDict(): Record<string, unknown> {
    const out: Record<string, unknown> = {};
    if (this.mode !== null) out.mode = this.mode;
    if (this.in_sidebar) out.in_sidebar = true;
    if (this.display_order !== 100) out.display_order = this.display_order;
    if (this.label !== null) out.label = this.label;
    if (this.icon !== null) out.icon = this.icon;
    if (this.description !== null) out.description = this.description;
    if (this.breadcrumb !== null) out.breadcrumb = [...this.breadcrumb];
    if (Object.keys(this.routes).length > 0) out.routes = { ...this.routes };
    if (Object.keys(this.permissions).length > 0) {
      const perms: Record<string, string[] | string> = {};
      for (const [k, v] of Object.entries(this.permissions)) {
        perms[k] = Array.isArray(v) ? [...v] : v;
      }
      out.permissions = perms;
    }
    if (this.note !== null) out.note = this.note;
    if (this.feature_flag !== null) out.feature_flag = this.feature_flag;
    return out;
  }

  /**
   * Resolve i18n label for the requested locale. Fallback order (mirrors
   * Python `resolve_label`):
   *   1. Exact locale match (e.g. 'pt-BR').
   *   2. Language-only fallback (e.g. 'pt' from 'pt-BR').
   *   3. 'en' (project default).
   *   4. null.
   * A plain-string label is returned as-is regardless of locale.
   */
  resolveLabel(locale = "en"): string | null {
    if (this.label === null) return null;
    if (typeof this.label === "string") return this.label;
    if (locale in this.label) return this.label[locale];
    if (locale.includes("-")) {
      const base = locale.split("-", 1)[0];
      if (base in this.label) return this.label[base];
    }
    return this.label.en ?? null;
  }
}
